### PR TITLE
Do not pass negative scores into function_score or script_score queries

### PR DESCRIPTION
**Pull request by msfroh**
In theory, Lucene scores should never go negative. To stop users from writing `function_score` and `script_score` queries that return negative values, we explicitly check their outputs and throw an exception when negative.

Unfortunately, due to a subtle, more complicated bug in multi_match queries, sometimes those might (incorrectly) return negative scores.

While that problem is also worth solving, we should protect function and script scoring from throwing an exception just for passing through a negative value that they had no hand in computing.

### Description
[Describe what this change achieves]

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/7860

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] ~~New functionality has been documented.~~
  - [ ] ~~New functionality has javadoc added~~
- [ ] ~~API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).~~
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] ~~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml [hunk 1]
```diff
@@ -482,3 +482,79 @@
...
+                "script_score": {
+                  "script": {
+                    "lang": "painless",
+                    "source": "_score"
+                  }
+                }
+              }]
+          explain: true
+  - match: { hits.total.value: 3 }
+  - match: { hits.hits.2._score: 0.0 }
```
Comment (mingshl) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml:527]: nice, so the functional score on a document will be at least zero, instead of negative. This will fix the score per field level. 

Wondering if this also fix the per field level score is negative? can we run profile api on yaml test and check perFieldSimilarity score is negative or zero?

Comment (msfroh) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml:527]: The logic focuses on the input score passed to a function or script, regardless of where that score came from.

Both before and after this change, we have checks that the score **returned** by a function or script is not negative (and throw an exception if it is). Unfortunately, this meant that a function/script that just passes a score through would throw an exception if the input score is negative.

So, for **any** case where a function or script gets an illegal negative value as input, this change defensively treats it as zero.

**Comment by msfroh** [comment]
Gradle check failed due to https://github.com/opensearch-project/OpenSearch/issues/13600. Retrying...

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml (hunk 1 above)
Comment (mingshl) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml:527]: awesome! thank you

File: server/src/main/java/org/opensearch/common/lucene/search/function/ScriptScoreFunction.java [hunk 2]
```diff
@@ -52,8 +52,14 @@
...
     static final class CannedScorer extends Scorable {
-        protected int docid;
-        protected float score;
+        private int docid;
+        private float score;
+
+        public void score(float subScore) {
+            // We check to make sure the script score function never makes a score negative, but we need to make
+            // sure the script score function does not receive negative input.
+            this.score = Math.max(0.0f, subScore);
```
Comment (reta) [inline server/src/main/java/org/opensearch/common/lucene/search/function/ScriptScoreFunction.java:61]: @msfroh don't we try to hide the problem instead of fixing the source of negative scoring? As you mentioned, we do catch that in some cases but it slips in other, why cannot we fix that?

Comment (msfroh) [reply server/src/main/java/org/opensearch/common/lucene/search/function/ScriptScoreFunction.java:61]: There was some discussion about this on https://github.com/opensearch-project/OpenSearch/issues/7860.

Essentially, the scoring logic in `cross_fields` match is pretty complicated. Actually, [the Elasticsearch documentation for `cross_fields`](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html#type-cross-fields) even contains the following warning: 

> The cross_fields type blends field statistics in a complex way that can be hard to interpret. The score combination can even be incorrect, in particular when some documents contain some of the search fields, but not all of them.

As @ylwu-amzn called out on the issue, any change we make to the `cross_fields` scoring formula could impact people's "regular" search results. We'd need to be extremely careful.

Meanwhile, this change just turns a query that was throwing an exception (which is what people have complained about) into one that doesn't.

We could possibly wrap the query returned from `MatchQueryBuilder` in a custom scorer that clamps negative values to zero. Functionally that's what happens if `tie_breaker` is 0. I would still include the changes in this PR as an extra layer of protection.

Comment (msfroh) [reply server/src/main/java/org/opensearch/common/lucene/search/function/ScriptScoreFunction.java:61]: @reta  -- I spent some more time today trying to prevent the negative score from bubbling up. So far as I can tell, this is the query that's returning the negative scores: https://github.com/opensearch-project/OpenSearch/blob/4a42150ba0cbf10022797c038f35faaea2f548f6/server/src/main/java/org/apache/lucene/queries/BlendedTermQuery.java#L410

Unfortunately, that just returns a Lucene `DisjunctionMaxQuery`, which normally does the correct thing. The problem arises because of the `adjustDF` and/or the `adjustTTF` methods. Let's pretend that Lucene term stats are something other than what they really are... yay!

I believe it's specifically `adjustTTF` method (called from the `blend` method) that breaks us, because it lowers the `sumTotalTermFreq` to the minimum of `sumTotalTermFreq` across all of the fields, which may be lower than any individual field's term's max `termFreq` (from another field). In theory, the shared `sumTotalTermFreq` should never be lower than the maximum of `termFreq` for any doc across all fields, but we don't have a good way of getting that value without iterating through all documents. I'd love to add a line just before [here](https://github.com/opensearch-project/OpenSearch/blob/4a42150ba0cbf10022797c038f35faaea2f548f6/server/src/main/java/org/apache/lucene/queries/BlendedTermQuery.java#L138) saying `minSumTTF = Math.max(minSumTTF, maxTermFreq)`.

tl;dr: Preventing the negative scores from getting into function/script scoring functions is hard. This `cross_field` scoring is a complicated mess. While it is the root cause, the immediate pain point is *not* the negative scores, but that the negative scores make perfectly valid function/script scoring throw an exception.

I'd like to commit this fix as-is to address the immediate problem. We can address the complicated `cross_field` logic down the road.

Comment (msfroh) [reply server/src/main/java/org/opensearch/common/lucene/search/function/ScriptScoreFunction.java:61]: When we enter the `blend()` method for the integ test I wrote, we start off with `color:red` with `df=3, sumTtf=3`, while `shape:red` has `df=1,sumTtf=1`. Also, `maxDoc` is 3.

1. We exit the first loop with `minSumTTF=2` and `max = 3`. We set `maxDoc` to `2` (since it's bigger than `minSumTTF`). 
2. [Later](https://github.com/opensearch-project/OpenSearch/blob/4a42150ba0cbf10022797c038f35faaea2f548f6/server/src/main/java/org/apache/lucene/queries/BlendedTermQuery.java#L163), we set `actualDf` to 2.

The relevant piece of the explain output is:

```
{
  "value" : -0.22314355,
  "description" : "idf, computed as log(1 + (N - n + 0.5) / (n + 0.5)) from:",
  "details" : [
    {
      "value" : 2,
      "description" : "n, number of documents containing term",
      "details" : [ ]
    },
    {
      "value" : 1,
      "description" : "N, total number of documents with field",
      "details" : [ ]
    }
  ]
},
```

That `n, number of documents containing the term` came from `actualDf`.

I think I can actually fix it by guaranteeing that `actualDf` is never greater than `N`, which I can get from `reader.getDocCount(terms[i].field())`.

So, after saying that fixing the underlying problem is too hard, I think I can fix the underlying problem instead. 😁

Comment (msfroh) [reply server/src/main/java/org/opensearch/common/lucene/search/function/ScriptScoreFunction.java:61]: Here we go: https://github.com/opensearch-project/OpenSearch/pull/13829

It may still be worth merging this PR as an extra layer of protection, but I'd need to update / remove the YAML REST test, because we no longer get a negative score that we force to zero. Alternatively, we could just close this PR, since the underlying problem is (probably) fixed.

Comment (reta) [reply server/src/main/java/org/opensearch/common/lucene/search/function/ScriptScoreFunction.java:61]: Thanks a lot for looking into this @msfroh (I learn quite a few things from your clear explanation)

**Comment by msfroh** [comment]
Canceling this one in favor of https://github.com/opensearch-project/OpenSearch/pull/13829

Thanks again @reta for pushing me to solve the underlying problem!
